src/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -62,11 +63,56 @@ public static bool TryGet(string headerName, out HeaderDescriptor descriptor)
             return true;
         }
 
+        // Returns false for invalid header name.
+        public static bool TryGet(ReadOnlySpan<byte> headerName, out HeaderDescriptor descriptor)
+        {
+            Debug.Assert(headerName.Length > 0);
+
+            KnownHeader knownHeader = KnownHeaders.TryGetKnownHeader(headerName);
+            if (knownHeader != null)
+            {
+                descriptor = new HeaderDescriptor(knownHeader);
+                return true;
+            }
+
+            if (!HttpRuleParser.IsToken(headerName))
+            {
+                descriptor = default(HeaderDescriptor);
+                return false;
+            }
+
+            descriptor = new HeaderDescriptor(ByteArrayHelpers.GetStringFromByteSpan(headerName));
+            return true;
+        }
+
         public HeaderDescriptor AsCustomHeader()
         {
             Debug.Assert(_knownHeader != null);
             Debug.Assert(_knownHeader.HeaderType != HttpHeaderType.Custom);
             return new HeaderDescriptor(_knownHeader.Name);
         }
+
+        public string GetHeaderValue(ReadOnlySpan<byte> headerValue)
+        {
+            if (headerValue.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            // If it's a known header value, use the known value instead of allocating a new string.
+            string[] knownValues = _knownHeader?.KnownValues;
+            if (knownValues != null)
```

Comment:
I'd have to look at what IL and asm gets generated here, but I wouldn't be surprised if there's an extra null check in the case where _knownHeader is null.  In that case, you're likely rechecking knownValues != null even though you know you just set it to null.